pkg/logs/sds/scanner.go

```diff
@@ -0,0 +1,295 @@
+				log.Error("Can't reconfigure SDS after having received standard rules:", rerr)
+				s.rawConfig = nil // we drop this configuration because it is unusable
+				if err == nil {
+					err = rerr
+				}
+			}
+		}
+		return err
+	case AgentConfig:
+		return s.reconfigureRules(order.Config)
+	}
+
+	return fmt.Errorf("Scanner.Reconfigure: Unknown order type: %v", order.Type)
+}
+
+// reconfigureStandardRules stores in-memory standard rules received through RC.
+// This is NOT reconfiguring the internal SDS scanner, call `reconfigureRules`
+// if you have to.
+// This method is NOT thread safe, the caller has to ensure the thread safety.
+func (s *Scanner) reconfigureStandardRules(rawConfig []byte) error {
+	if rawConfig == nil {
+		return fmt.Errorf("Invalid nil raw configuration for standard rules")
+	}
+
+	var standardRules StandardRulesConfig
+	if err := json.Unmarshal(rawConfig, &standardRules); err != nil {
+		return fmt.Errorf("Can't unmarshal raw configuration: %v", err)
+	}
+
+	s.standardRules = standardRules
+	log.Info("Reconfigured SDS standard rules.")
+	return nil
+}
+
+// reconfigureRules reconfigures the internal SDS scanner using the in-memory
+// standard rules. Could possibly delete and recreate the internal SDS scanner if
+// necessary.
+// This method is NOT thread safe, caller has to ensure the thread safety.
+func (s *Scanner) reconfigureRules(rawConfig []byte) error {
+	if rawConfig == nil {
+		telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationError, 1, []string{"error:null_config"})
+		return fmt.Errorf("Invalid nil raw configuration received for user configuration")
+	}
+
+	if s.standardRules.Rules == nil {
+		telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationError, 1, []string{"error:no_standard_rules"})
+		// store it for the next try
+		s.rawConfig = rawConfig
+		log.Info("Received an user configuration before receiving the SDS standard rules.")
+		return nil
+	}
+
+	var config RulesConfig
+	if err := json.Unmarshal(rawConfig, &config); err != nil {
+		telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationError, 1, []string{"error:invalid_json"})
+		return fmt.Errorf("Can't unmarshal raw configuration: %v", err)
+	}
+
+	// ignore disabled rules
+	config = config.OnlyEnabled()
+
+	// if we received an empty array of rules or all rules disabled, interprets this as "stop SDS".
+	if len(config.Rules) == 0 {
+		telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationSuccess, 1, []string{})
+		log.Info("Received an empty configuration, stopping the SDS scanner.")
+		// destroy the old scanner
+		if s.Scanner != nil {
+			s.Scanner.Delete()
+			s.Scanner = nil
+			s.rawConfig = rawConfig
+			s.configuredRules = nil
+			return nil
+		}
+		return nil
+	}
+
+	// prepare the scanner rules
+	var sdsRules []sds.Rule
+	for _, userRule := range config.Rules {
+		// read the rule in the standard rules
+		standardRule := s.standardRules.GetByID(userRule.Definition.StandardRuleID)
```

Comment:
Hundreds, worse case thousands, I don't think it matters much since it's only when reconfiguring. But still it won't be a big effort to use a map instead, I'll change this. Good catch thank you.

Reply:
Addressed.